args/core/ecs/ecsregistry.cpp

```diff
@@ -1,71 +1,206 @@
+		async::readonly_guard guard(m_familyLock);
+
+		if (!m_families.contains(componentTypeId))
 			throw args_unknown_component_error;
 
-		return families[componentTypeId];
+		return m_families[componentTypeId];
 	}
 
 	inline component_handle_base EcsRegistry::getComponent(id_type entityId, id_type componentTypeId)
 	{
-		if (getEntity(entityId).has_component(componentTypeId))
-			return component_handle_base(entityId, *this);
-		return component_handle_base(invalid_id, *this);
+		if (!validateEntity(entityId))
+			throw args_entity_not_found_error;
+
+		return component_handle_base(entityId, *this);
 	}
 
 	inline component_handle_base EcsRegistry::createComponent(id_type entityId, id_type componentTypeId)
 	{
-		entity& entity = getEntity(entityId);
+		if (!validateEntity(entityId))
+			throw args_entity_not_found_error;
 
 		getFamily(componentTypeId)->create_component(entityId);
 
-		entity.m_components.insert(componentTypeId, componentTypeId);
+		{
+			async::readonly_guard guard(m_entityDataLock);
+			m_entityData[entityId].components.insert(componentTypeId);
+		}
+
+		m_queryRegistry.evaluateEntityChange(entityId, componentTypeId, true);
 
 		return component_handle_base(entityId, *this);
 	}
 
-	void EcsRegistry::destroyComponent(id_type entityId, id_type componentTypeId)
+	inline void EcsRegistry::destroyComponent(id_type entityId, id_type componentTypeId)
 	{
-		entity& entity = getEntity(entityId);
+		if (!validateEntity(entityId))
+			throw args_entity_not_found_error;
 
 		getFamily(componentTypeId)->destroy_component(entityId);
 
-		entity.m_components.erase(componentTypeId);
+		{
+			async::readonly_guard guard(m_entityDataLock);
+			m_entityData[entityId].components.erase(componentTypeId);
+		}
+
+		m_queryRegistry.evaluateEntityChange(entityId, componentTypeId, true);
+	}
+
+	A_NODISCARD inline bool EcsRegistry::validateEntity(id_type entityId)
+	{
+		async::readonly_guard guard(m_entityLock);
+		return entityId && m_entities.contains(entityId);
 	}
 
-	inline entity& EcsRegistry::createEntity()
+	inline entity_handle EcsRegistry::createEntity()
 	{
-		id_type id = lastEntityId++;
+		id_type id = m_nextEntityId++;
 
-		if (entities.count(id))
+		if (validateEntity(id))
 			throw args_entity_exists_error;
 
-		entities[id] = new entity(id, *this);
+		{
+			async::readwrite_guard guard(m_entityDataLock);
+			m_entityData[id] = {};
+		}
 
-		return *entities[id];
+		async::readwrite_guard guard(m_entityLock);
+		m_entities.emplace(id, id, this);
+
+		return m_entities[id];
+	}
+
+	inline void EcsRegistry::destroyEntity(id_type entityId, bool recurse)
+	{
+		if (!validateEntity(entityId))
+			throw args_entity_not_found_error;
+
+		entity_data* data = nullptr;
+
+		{
+			async::readonly_guard guard(m_entityDataLock);
+			data = &m_entityData[entityId];
+		}
+
+		{
+			async::readonly_guard guard(m_familyLock); // Possibly deadlocks? Will also lock atomic_sparse_map::m_container_lock for the family.
```

Comment:
if it is a potential deadlock , please mark with TODO(\<assignee\>): to figure out